### PR TITLE
Add Serial API post

### LINKED ISSUE #3194: content: Serial [2020-07-10 HARD]
* [draft](https://docs.google.com/document/d/1i5RmYodf0CfMEWYunje6Ei3yPG3aRS1pl39YJ7uVA1k/edit)
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/DLAn-HDih3M)
* next step: @beaufortfrancois to address feedback

### PR DESCRIPTION
**Pull request by beaufortfrancois**
This PR adds the article dedicated to the Serial API and also updates the associated https://web.dev/fugu-status page.

The draft has been reviewed by @jpmedley and @reillyeon.

FYI @petele 

Live preview:
https://deploy-preview-3688--web-dev-staging.netlify.app/serial/
https://deploy-preview-3688--web-dev-staging.netlify.app/fugu-status/#origin-trial
 
Fixes #3194

### CODE REVIEW

**Comment by beaufortfrancois** [comment]
@reillyeon I'm thinking we should mention in this article changes you've made recently:
 - discarding buffered data is now possible by calling cancel()  and abort() on the readable and writable attributes of the SerialPort object.
- waiting for buffered data to be written is now possible by calling close() on the writable attribute of the SerialPort object

WDYT?

File: src/site/content/en/blog/serial/index.md [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+reader and writer.
+
+```js
+await port.close();
+```
+
+However, when continuously reading data from a serial device using a loop,
+`port.readable` will always be locked until it encounters an error. In this
+case, calling `reader.cancel()` will force `reader.read()` to resolve
+immediately with `{ value: undefined, done: true }` and therefore allowing loop
```
Comment (jpmedley) [inline src/site/content/en/blog/serial/index.md:315]:
```suggestion
immediately with `{ value: undefined, done: true }` and therefore allowing the loop
```

File: src/site/content/en/blog/serial/index.md [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+  // TODO: Remove |event.port| from the UI.
+  // If the serial port was opened, a stream error would be observed as well.
+});
+```
+
+### Handle signals {: #signals }
+
+After establishing the serial port connection, you can explicitly query and set
+signals exposed by the serial port for device detection and flow control. These
+signals are defined as boolean values. For example, some devices like Arduino
```
Comment (jpmedley) [inline src/site/content/en/blog/serial/index.md:400]:
```suggestion
signals are defined as boolean values. For example, some devices such as Arduino
```

File: src/site/content/en/blog/serial/index.md [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+parse the incoming stream and return parsed data. The transform stream sits
+between the serial device and the read loop that is consuming the stream. It can
+apply an arbitrary transform before the data is consumed. Think of it like an
+assembly line: as a widget comes down the line, each step in the line modifies
+the widget, so that by the time it gets to its final destination, it's a fully
+functioning widget.
+
+<figure class="w-figure">
+  <img src="./aeroplane-factory.jpg" alt="Photo of an aeroplane factory">
+  <figcaption class="w-figcaption">World War 2 Castle Bromwich Aeroplane Factory</figcaption>
```
Comment (jpmedley) [inline src/site/content/en/blog/serial/index.md:443]:
```suggestion
  <figcaption class="w-figcaption">World War II Castle Bromwich Aeroplane Factory</figcaption>
```

Comment (jpmedley) [reply src/site/content/en/blog/serial/index.md:443]: I've never seen it done with anything other than Roman numerals in English.

File: src/site/content/en/blog/serial/index.md [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+For debugging serial device communication issues, use the `tee()` method of
+`port.readable` to split the streams going to or from the serial device. The two
+streams created can be consumed independently and this allows you to print one
+to the console for inspection.
+
+```js
+const [appReadable, devReadable] = port.readable.tee();
+
+// TODO: Update UI with incoming data from appReadable.
+// TODO: Log incoming data in JS console for inspection from devReadable.
```
Comment (jpmedley) [inline src/site/content/en/blog/serial/index.md:497]: I'm thinking "TODO" should be removed. People might think we neglected to finish the code example.

File: src/site/content/en/blog/serial/index.md [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+## Polyfill {: #polyfill }
+
+On Android, support for USB-based serial ports is possible using the WebUSB API
+and the [Serial API polyfill]. This polyfill is limited to hardware and
+platforms where the device is accessible via the WebUSB API because it has not
+been claimed by a built-in device driver.
+
+## Security and privacy {: #security-privacy }
+
+The Chrome team has designed and implemented the Serial API using the core
```
Comment (jpmedley) [inline src/site/content/en/blog/serial/index.md:514]:
```suggestion
The spec authors designed and implemented the Serial API using the core
```

File: src/site/content/en/blog/serial/index.md (hunk 4 above)
Comment (beaufortfrancois) [reply src/site/content/en/blog/serial/index.md]: Fixed.

File: src/site/content/en/blog/serial/index.md (hunk 5 above)
Comment (beaufortfrancois) [reply src/site/content/en/blog/serial/index.md]: Fixed. But what is the rationale?

**Comment by reillyeon** [comment]
> @reillyeon I'm thinking we should mention in this article changes you've made recently:
> 
> * discarding buffered data is now possible by calling cancel()  and abort() on the readable and writable attributes of the SerialPort object.
> * waiting for buffered data to be written is now possible by calling close() on the writable attribute of the SerialPort object
> 
> WDYT?

These are niche capabilities that only apply to a small portion of applications. It's important that developers know it is available but we should avoid encouraging developers to use it all the time.

Should we add a note to the article that some of the names have recently changed and developer should be sure they are running Chrome 86 or later?

**Comment by beaufortfrancois** [comment]
> Should we add a note to the article that some of the names have recently changed and developer should be sure they are running Chrome 86 or later?

@jpmedley Would a note like "This article reflects the Serial API as implemented in Chrome 86 and later." look good to you?

image

**Comment by reillyeon** [comment]
Maybe add, "Some property names have changed from previous versions."

File: src/site/content/en/blog/serial/index.md (hunk 5 above)
Comment (jpmedley) [reply src/site/content/en/blog/serial/index.md]: It seems that Reilly listed spec authors in a way that credits everyone who contributed to the spec repo in some way. Not all of them are Google employees.

**Comment by jpmedley** [comment]
@beaufortfrancois LGTM provided you add Reilly's line.

**Comment by beaufortfrancois** [comment]
@jpmedley Can you click the "approve request" button and merge this PR?

**Comment by kaycebasques** [comment]
Waiting on CI to pass (might be in a weird state)

image

**Comment by kaycebasques** [comment]
@beaufortfrancois please try rebasing this PR to be up-to-date with `master`

**Pull request merged**